modules/renter/contractor/allowance.go

```diff
@@ -0,0 +1,161 @@
+	}
+
+	c.mu.RLock()
+	// gather contracts to renew
+	var renewSet []modules.RenterContract
+	for _, contract := range c.contracts {
+		renewSet = append(renewSet, contract)
+	}
+
+	// calculate new endHeight; if the period has not changed, the endHeight
+	// should not change either
+	endHeight := c.blockHeight + a.Period
+	if a.Period == c.allowance.Period {
+		for _, contract := range c.contracts {
+			endHeight = contract.EndHeight()
+			break
+		}
+	}
+	c.mu.RUnlock()
+
+	// renew existing contracts with new allowance parameters
+	newContracts := make(map[types.FileContractID]modules.RenterContract)
+	for _, contract := range renewSet {
+		newContract, err := c.managedRenew(contract, numSectors, endHeight)
+		if err != nil {
+			c.log.Printf("WARN: failed to renew contract with %v; a new contract will be formed in its place", contract.NetAddress)
+			remaining++
+			continue
+		}
+		newContracts[newContract.ID] = newContract
+		if len(newContracts) >= int(a.Hosts) {
+			break
+		}
+	}
+
+	// if we did not renew enough contracts, form new ones
+	if remaining > 0 {
+		formed, err := c.managedFormContracts(remaining, numSectors, endHeight)
+		if err != nil {
+			return err
+		}
+		for _, contract := range formed {
+			newContracts[contract.ID] = contract
+		}
+	}
+
+	// if we weren't able to form anything, return an error
+	if len(newContracts) == 0 {
+		return errors.New("unable to form or renew any contracts")
+	}
+
+	// Set the allowance and replace the contract set
+	c.mu.Lock()
+	c.allowance = a
+	c.contracts = newContracts
+	// update metrics
+	var spending types.Currency
+	for _, contract := range c.contracts {
+		spending = spending.Add(contract.RenterFunds())
+	}
+	c.financialMetrics.ContractSpending = spending
+	err = c.saveSync()
+	c.mu.Unlock()
+
+	return err
+}
+
+// managedFormAllowanceContracts handles the special case where no contracts
+// need to be renewed when setting the allowance.
+func (c *Contractor) managedFormAllowanceContracts(n int, numSectors uint64, a modules.Allowance) error {
+	if n <= 0 {
+		return nil
+	}
+
+	// if we're forming contracts but not renewing, the new contracts should
+	// have the same endHeight as the existing ones. Otherwise, the endHeight
+	// should be a full period.
+	c.mu.RLock()
+	endHeight := c.blockHeight + a.Period
+	for _, contract := range c.contracts {
+		endHeight = contract.EndHeight()
+		break
+	}
+	c.mu.RUnlock()
+
+	// form the contracts
+	formed, err := c.managedFormContracts(n, numSectors, endHeight)
+	if err != nil {
+		return err
+	}
+
+	// Set the allowance and replace the contract set
+	c.mu.Lock()
+	c.allowance = a
+	for _, contract := range formed {
+		c.contracts[contract.ID] = contract
+	}
+	// update metrics
+	var spending types.Currency
+	for _, contract := range c.contracts {
+		spending = spending.Add(contract.RenterFunds())
+	}
+	c.financialMetrics.ContractSpending = spending
+	err = c.saveSync()
+	c.mu.Unlock()
```

Comment:
So it looks like the atomic swap thing is happening in this code. And the eventual plan (post 1.0) is to have some sort of saved field where contracts that were successfully formed but not swapped in get restored and you finish whatever didn't get finished the first time?